[gz] optical flow support

dakejahl opened the pull request:
Opening this draft PR for review and testing

gz models branch
https://github.com/dakejahl/PX4-gazebo-models/tree/pr-rangefinder

dirksavage88 commented:
There is less drift on the textured world:

[flowtestgz.zip](https://github.com/user-attachments/files/16393204/flowtestgz.zip)

dakejahl commented:
I didn't realize gazebo sitl already had flow so I am going to update this to use the OpticalFlow repo
https://github.com/PX4/PX4-SITL_gazebo-classic/blob/main/src/gazebo_opticalflow_plugin.cpp
https://github.com/PX4/PX4-OpticalFlow

dakejahl commented:
@Jaeyoung-Lim I added the PX4/OpticalFlow in my latest commit. The performance is much worse, it just toilet bowls and wanders. Any ideas?

FYI for testing ensure EKF2_OF_CTRL is enabled and SIM_GPS_USED is set to 0 to ensure only flow is used for position hold.

Jaeyoung-Lim commented:
@dakejahl I would check the camera orientation if it toilet bowls

dakejahl commented:
@Jaeyoung-Lim The orientation looks correct. I think the issue is the noise, I was seeing this in my original implementation when poor quality keypoints are used in the flow average.
![image](https://github.com/user-attachments/assets/fa4d2f66-e5c5-4d61-8202-125744fd9ccc)

dakejahl commented:
> @Jaeyoung-Lim I added the PX4/OpticalFlow in my latest commit. The performance is much worse, it just toilet bowls and wanders. Any ideas?

The issue was `quality` was -1 and wrapping to 255. I added a fix and it is a lot better now.

dirksavage88 commented:
It's working. I think there was an issue with the libOpticalFlow in the cmakelist, it was working after I placed it further down in the camera folder's cmakelist, after add_library

dirksavage88 commented on src/modules/simulation/gz_bridge/camera/CMakeLists.txt:
```diff
@@ -0,0 +1,81 @@
+message(WARNING "OpticalFlow install dir: ${install_dir}")
+
+set(OpticalFlow_INCLUDE_DIRS ${install_dir}/include)
+set(OpticalFlow_LIBS ${install_dir}/lib/libOpticalFlow.so)
```
this line resulted in an error, cmake tries to find it before it even compiles the shared object?

dakejahl commented:
I'm not sure why but sometimes when sim is launched PX4 reports
```
WARN  [health_and_arming_checks] Preflight Fail: ekf2 missing data
WARN  [health_and_arming_checks] Preflight Fail: No connection to the ground control station	
WARN  [health_and_arming_checks] Preflight Fail: ekf2 missing data
```
And it never recovers

Jaeyoung-Lim commented on src/modules/simulation/gz_bridge/camera/gz_camera.cpp:
```diff
@@ -0,0 +1,35 @@
+	int quality = _optical_flow->calcFlow(gray_image.data, sim_time, integration_time, flow_x, flow_y);
+
+	return quality;
+}
```
I think we should add a sensor plugin for an optical flow sensor, not calculate optical flow inside the gzbridge. Then it is not a bridge anymore

dakejahl replied:
hmm yeah I was thinking about that. I need to figure out how to create a gz plugin that is launched with sim since it doesn't exist yet for gz
